Respect empty `expiryDate` value in server

nfebe opened the pull request:
If `expireDate` is an empty string and not `null` then the server should not set a default.


Resolves : https://github.com/nextcloud/server/issues/44219

nfebe committed:
fix(shareManager): Respect empty `expireDate` in server

If `expireDate` is an empty string and not `null` then the server should not set a default.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

Fenn-CS commented:
Related changes : https://github.com/nextcloud/server/pull/44912

nfebe committed:
refactor(shareManager): Use constructor property promotion

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

Fenn-CS commented:
/backport to stable29

Fenn-CS commented:
/backport to stable28

Fenn-CS commented:
/backport to stable27

nickvergessen commented on lib/private/Share20/Manager.php:
```diff
@@ -487,25 +417,6 @@ protected function validateExpirationDateLink(IShare $share) {
 			}
 		}
 
-		// If expiredate is empty set a default one if there is a default
```
but you are removing the default functionality?

Fenn-CS replied:
Yes :)... Since the users now have the opportunity to modify this settings before a share is created. `i.e` the user can see on the share creation menu, that an expiry date is required and can either allow the default or remove if not enforced.

Leaving this default functionality means, we always force shares to have expiry dates even when that's not the users intention.

nickvergessen replied:
The problem is that currently existing clients and app might rely on the default and not populate it via the UI or send it with each request.
So this is a breaking change that is impossible to discover and will lead to a lot of shares not expiring although they should.

Fenn-CS replied:
I see. This breaking change is inevitable if we would solve https://github.com/nextcloud/server/issues/44219 though.

I am not sure but perhaps I could introduce checks for clients/app calls.

Another though is that `Manager` class is only part of the private API and maybe all we need to do is make sure the apps we maintain that uses the non public share interface are updated?

nickvergessen replied:
> Another though is that Manager class is only part of the private API and maybe all we need to do is make sure the apps we maintain that uses the non public share interface are updated?

I meant apps like Android, iOS and our Desktop Client:
- https://github.com/nextcloud/server/blob/490f545006024ad982c801b5c986ff8d03826821/apps/files_sharing/lib/Controller/ShareAPIController.php#L574
- https://github.com/nextcloud/server/blob/490f545006024ad982c801b5c986ff8d03826821/apps/files_sharing/lib/Controller/ShareAPIController.php#L648-L656

That being said, the update also supports it, but basically you need to set it to `''` (empty string) not null when updating a share and removing a expiration date.

Seems to work when updating, just not when creating, as the check is always executed then.
So instead of tampering with `validateExpirationDateInternal` like this, I would try to get the information into the share object that no expirationDate was sent and in https://github.com/nextcloud/server/blob/ec5133b739eabc76271789504b4dbb91a534f552/lib/private/Share20/Manager.php#L423 instead of only checking `$expirationDate === null` you also check whether it was sent with the empty string, and only set the default date when it was not sent as empty string.

nfebe reviewed (commented):
Backtracked on the initial radical and breaking approach where, we allow defaults to be set entire by the client.

Propose new approach: 

- If client sends empty string, then do not set any defaults
- Set defaults it no `expireDate` is sent
- Set default if `expireDate` sent is null

Fenn-CS commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -646,12 +646,16 @@ public function createShare(
+					throw new OCSNotFoundException($this->l->t('Invalid date, date format must be YYYY-MM-DD'));
+				}
+			} else {
+				$share->setExpirationDate('');
```
Since, `null` is the basis for setting default expiration date in the share manager for new shares, we just can't set this to null here.

Fenn-CS replied:
```suggestion
				$share->setExpirationDate(null);
```

Since, `null` is the basis for setting default expiration date in the share manager for new shares, we just can't set this to null here.


But can be done much later in `validateExpirationDateInternal` and `validateExpirationDateLink`

provokateurin commented on lib/public/Share/IShare.php:
```diff
@@ -385,11 +385,11 @@ public function getNote();
 	 * Set the expiration date
 	 *
-	 * @param null|\DateTime $expireDate
+	 * @param \DateTime|string|null $expireDate
```
```suggestion
	 * @param \DateTime|''|null $expireDate
```
To avoid any other string than the empty string. I'm questioning a bit if this couldn't be solved differently because two "null" values are not great

nickvergessen replied:
Which is why i would have added another prop for this. Also helps to improve the enforcement handling above

nickvergessen commented on lib/private/Share20/Manager.php:
```diff
@@ -389,6 +337,13 @@ protected function validateExpirationDateInternal(IShare $share) {
 		$expirationDate = $share->getExpirationDate();
 
+		// If $expirationDate is an empty string, then the client unset expiration  date
+		// Hence no validation required
```
This is not correct. If the enforcement is enabled you still have to.

nickvergessen commented on lib/private/Share20/Manager.php:
```diff
@@ -475,6 +430,13 @@ protected function validateExpirationDateInternal(IShare $share) {
 		$expirationDate = $share->getExpirationDate();
 
+		// If $expirationDate is an empty string, then the client unset expiration  date
+		// Hence no validation required
```
Same here

come-nc commented on lib/private/Share20/Share.php:
```diff
@@ -41,61 +41,34 @@
-	/** @var string */
-	private $sharedWithAvatar;
-	/** @var string */
-	private $sharedBy;
-	/** @var string */
-	private $shareOwner;
-	/** @var int */
-	private $permissions;
-	/** @var IAttributes */
-	private $attributes;
-	/** @var int */
-	private $status;
-	/** @var string */
-	private $note = '';
-	/** @var \DateTime */
-	private $expireDate;
-	/** @var string */
-	private $password;
+	private ?string $id = null;
+	private ?string $providerId = null;
+	private ?Node $node = null;
+	private ?int $fileId = null;
+	private ?string $nodeType = null;
+	private ?int $shareType = null;
+	private ?string $sharedWith = null;
+	private ?string $sharedWithDisplayName = null;
+	private ?string $sharedWithAvatar = null;
+	private ?string $sharedBy = null;
+	private ?string $shareOwner = null;
+	private ?int $permissions = null;
+	private ?IAttributes $attributes = null;
+	private ?int $status = null;
+	private ?string $note = null;
+	private ?\DateTime $expireDate = null;
+	private bool $overwriteFalsyExpirationDate = true;
+	private ?string $password = null;
 	private ?\DateTimeInterface $passwordExpirationTime = null;
-	/** @var bool */
-	private $sendPasswordByTalk = false;
-	/** @var string */
```
You’re gonna get trouble with all this strong typing. I tried it in the psalm update PR and typing in this class is inconsistent between props and methods. Main problem is the class can be used to represent an incomplete share while some code path expect it to be complete.

I’d advice not touching the typing or do it in a separate PR which fixes the problems.

Fenn-CS replied:
Moved to : https://github.com/nextcloud/server/pull/45156

come-nc reviewed (commented):
I have to say I do not understand why the share object has to carry the information of whether a default date was used, it should simply contain the date. So this new method in IShare is not logical to me.

Code around this sounds quite convoluted.

come-nc commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -646,12 +646,17 @@ public function createShare(
+					$expireDateTime = $this->parseDate($expireDate);
+					$share->setExpirationDate($expireDateTime);
+				} catch (\Exception $e) {
+					throw new OCSNotFoundException($this->l->t('Invalid date, date format must be YYYY-MM-DD'));
```
`	 * @throws OCSNotFoundException Creating the share failed
`
Apparently not :thinking: 
This API is really weird.

Fenn-CS replied:
To avoid more back and forth for something unrelated moved to : https://github.com/nextcloud/server/pull/45157

come-nc commented:
> I have to say I do not understand why the share object has to carry the information of whether a default date was used, it should simply contain the date. So this new method in IShare is not logical to me.
> 
> Code around this sounds quite convoluted.

Ok I’m starting to understand, the share object is used both to convey the parameter for creation and the final created object, this is quite bad.
So I do not see a better way than what you did for storing the info without a bigger refactor.
I do not understand the name `setOverwriteFalsyExpirationDate` though.
Why not `setNoExpirationDate(bool)` ?

nickvergessen commented on lib/public/Share/IShare.php:
```diff
@@ -385,20 +385,38 @@ public function getNote();
+	 * @return \OCP\Share\IShare The modified object
+	 * @since 27.0.0
+	 */
+	public function setOverwriteFalsyExpirationDate(bool $overwriteFalsyExpirationDate);
```
```suggestion
	public function setNoExpirationDate(bool $noExpirationDate);
```
etc

Fenn-CS commented:
> I do not understand the name setOverwriteFalsyExpirationDate though.
> Why not setNoExpirationDate(bool) ?

The method name `setOverwriteFalsyExpirationDate` was chosen for its explicitness in indicating that the method controls whether to overwrite falsy values, such as `null` or an empty string (`''`), in the `expireDate` field. 

Because, that is exactly what the method and its corresponding property is currently used for in the code, we use it to decide if we want to **overwrite** the value of `$expireDate` when it has *falsy* values (`null`, `''`, `0'`, etc) ...

However, following your suggestion, I have changed the name to `setNoExpirationDate(bool)` as it might be easier especially for non-native English speakers to grab.

Fenn-CS commented on lib/private/Share20/Manager.php:
```diff
@@ -474,6 +429,13 @@ protected function validateExpirationDateInternal(IShare $share) {
 	 */
 	protected function validateExpirationDateLink(IShare $share) {
```
Yes the method is in use.

Fenn-CS commented on lib/private/Share20/Manager.php:
```diff
@@ -474,6 +429,13 @@ protected function validateExpirationDateInternal(IShare $share) {
+
+		// If $expirationDate is falsy, overwrite flag is false, no expiration is set
+		if(empty($expirationDate) && !$share->getOverwriteFalsyExpirationDate() && !$isEnforced) {
+			$share->setExpirationDate(null);
```
Yes because `''` (empty string) value would lead to `DBAL` exceptions when trying to convert a string into value that should be stored on a `nullable` `datetime` field.

come-nc replied:
But expirationDate is null or DateTime it cannot be `''` in the share object.

Fenn-CS commented on lib/private/Share20/Manager.php:
```diff
@@ -409,17 +375,6 @@ protected function validateExpirationDateInternal(IShare $share) {
-		}
 		if ($fullId === null && $expirationDate === null && $defaultExpireDate) {
 			$expirationDate = new \DateTime('now', $this->dateTimeZone->getTimeZone());
 			$expirationDate->setTime(0, 0, 0);
```
@come-nc  not clear here please...

Fenn-CS replied:
Figured.

nickvergessen commented:
Would really be good to add a new test into https://github.com/nextcloud/server/tree/058b69eef02be9bfa77e9b6a1d12d15f29ac6b0c/build/integration/sharing_features which proves the thing you fixed, so it does not get reverted at some point and no test fails :P

come-nc commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -549,7 +549,8 @@ public function deleteShare(string $id): DataResponse {
 	 * @param string $password Password for the share
 	 * @param string|null $sendPasswordByTalk Send the password for the share over Talk
-	 * @param string $expireDate Expiry date of the share using user timezone at 00:00. It means date in UTC timezone will be used.
+	 * @param ?string $expireDate The expiry date of the share in the user's timezone (UTC) at 00:00.
```
```suggestion
	 * @param ?string $expireDate The expiry date of the share in the user's timezone at 00:00.
```
The user’s timezone is not always UTC

come-nc commented:
I agree that we need tests for this.

blizzz commented:
Is it possible to get some momentum into this PR? With regard to Red CI and the suggested test?

nfebe committed:
test(Sharing): Integration test for no expiration set date for share

- Verify that explicitly sending empty `expireDate` param to server overwrite default
and sets not expiry date, if non is enforced.

- Update tests to avoid converting empty string to date.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

Fenn-CS commented on build/integration/sharing_features/sharing-v1.feature:
```diff
@@ -229,6 +229,23 @@ Feature: sharing
       | url | AN_URL |
       | mimetype | httpd/unix-directory |
 
+  Scenario: Creating a new share with expiration date removed, when default expiration is set
```
Would fix up commit to main changes once tests are approved

nfebe reviewed (commented):
WIP commit present.

Fenn-CS commented on lib/private/Share20/Manager.php:
```diff
@@ -379,28 +307,53 @@ protected function validateExpirationDateInternal(IShare $share) {
+
+		// If $expirationDate is falsy, noExpirationDate is true and expiration not enforced
+		// Then skip expiration date validation as null is accepted
+		if(!(empty($expirationDate) && $share->getNoExpirationDate() && !$isEnforced)) {
```
Does not make a difference, would update.

```suggestion
		if(!($share->getNoExpirationDate() && !$isEnforced)) {
```

nfebe reviewed (commented):
If `noExpirationDate` is set to true and no enforcement no validation is needed irresepective of the value of `expirationDate`

Fenn-CS commented on build/integration/sharing_features/sharing-v1.feature:
```diff
@@ -229,6 +229,24 @@ Feature: sharing
+    Then the OCS status code should be "100"
+    And the HTTP status code should be "200"
+    And Share fields of last share match with
+      | expiration ||
```
@artonge In case you are wondering this is what the test suite is complaining about.

Context:

- We updated the backend to not set expiry date when an empty value is sent
- So when the expiry is created it returns expiration that should be nullable or an empty space.
- So next step is to try `| null |` currently I am passing `||` and not sure how Gherkin would parse that.

come-nc commented on lib/public/Share/IShare.php:
```diff
@@ -385,20 +385,38 @@ public function getNote();
+	 *
+	 * @param bool $noExpirationDate
+	 * @return \OCP\Share\IShare The modified object
+	 * @since 27.0.0
```
```suggestion
	 * @since 29.0.0
```
And then adapted in backports

come-nc replied:
(same for  `getNoExpirationDate`)

ShGKme commented:
This breaks uploading in Talk which expects `$expireDate` to be `string`

blizzz commented:
> This breaks uploading in Talk which expects `$expireDate` to be `string`

Also on 27?

Antreesy commented:
> Also on 27?

Confirm for stable27
![image](https://github.com/nextcloud/server/assets/93392545/54dbc357-e214-41f4-8777-731c8fa8e6d2)

nickvergessen commented on lib/public/Share/IShare.php:
```diff
@@ -385,20 +385,38 @@ public function getNote();
+	 * @return bool
+	 * @since 30.0.0
+	 */
+	public function getNoExpirationDate();
```
@Fenn-CS Please document in https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_30.html#id1

(all changes in lib/public/)

Fenn-CS replied:
https://github.com/nextcloud/documentation/pull/12051
